pandas/_libs/lib.pxd

```diff
@@ -1 +1 @@
-cdef bint c_is_list_like(object, bint)
+cdef bint c_is_list_like(object, bint) except -1
```

Comment:
Hmm don't think a bint can ever return -1. The only thing I think you could except here is NULL but would really have to refactor to do that

Is there a particular issue you are trying to solve?

Reply:
Thanks for reviewing.

The ignored exception led to a segfault in https://github.com/rapidsai/cudf that I'm unable to reproduce at the moment since we've since refactored our code a bit -- apologies.

However, we can demonstrate the issue by raising artificially within `c_is_list_like` -- I believe it's valid to assume that Python exceptions can originate from within this function.

1. Without the `except -1`:

```cython
cdef inline bint c_is_list_like(object obj, bint allow_sets):
    raise ZeroDivisionError()
    return (
        isinstance(obj, abc.Iterable)
        # we do not count strings/unicode/bytes as list-like
        and not isinstance(obj, (str, bytes))
        # exclude zero-dimensional numpy arrays, effectively scalars
        and not (util.is_array(obj) and obj.ndim == 0)
        # exclude sets if allow_sets is False
        and not (allow_sets is False and isinstance(obj, abc.Set))
    )
```


```python
>>> import pandas as pd
>>> pd.api.types.is_list_like([])
ZeroDivisionError
Exception ignored in: 'pandas._libs.lib.c_is_list_like'
ZeroDivisionError:
False
```

2. With the `except -1`:

```
cdef inline bint c_is_list_like(object obj, bint allow_sets) except -1:                                                                                                                                             raise ZeroDivisionError()
    return (                                                                                                                                                                                                            isinstance(obj, abc.Iterable)
        # we do not count strings/unicode/bytes as list-like                                                                                                                                                            and not isinstance(obj, (str, bytes))                                                                                                                                                                           # exclude zero-dimensional numpy arrays, effectively scalars
        and not (util.is_array(obj) and obj.ndim == 0)                                                                                                                                                                  # exclude sets if allow_sets is False                                                                                                                                                                           and not (allow_sets is False and isinstance(obj, abc.Set))
    )
```

```python
>>> import pandas as pd
>>>pd.api.types.is_list_like([])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/_libs/lib.pyx", line 985, in pandas._libs.lib.is_list_like
    return c_is_list_like(obj, allow_sets)
  File "pandas/_libs/lib.pyx", line 989, in pandas._libs.lib.c_is_list_like
    raise ZeroDivisionError()
ZeroDivisionError
```

(`-1` seems to be a [valid value](https://github.com/cython/cython/blob/master/tests/run/bint.pyx#L16) for `bint`)

Reply:
> The ignored exception led to a segfault

Any idea what value was being passed?  Its hard to imagine what could cause this to raise unless you specifically rigged it to make isinstance checks fail

Reply:
Aha - so after digging up some old code, it looks the source of the problem was a `RecursionError`, which happened to be thrown within `isinstance` here.

Reply:
Minimal repro:

```
In [1]: import pandas as pd

In [2]: def foo():
   ...:     pd.api.types.is_list_like("test")
   ...:     foo()
```

Reply:
-1 is appropriate here as it signals cython that there *maybe* is an exception to check.